### DISCUSSION TITLE
OpenFAST as a function

### DISCUSSION THREAD

**Discussion by HiHoGit** [OpenFAST/openfast#1279] [Q&A]
Dear all,
I need to run a fully nonlinear version of OpenFAST as a function in order to get some outputs when I have some others depending on conditions. for example, when I have tower top acceleration, be able to calculate or estimate the deflection or shape function. Does OpenFAST have such a possibility?

**Comment by jjonkman** [comment]
Dear @HiHoGit,

Tower-top acceleration is an output from, not an an input to, OpenFAST.  So, "no", such functionality is not directly available in OpenFAST.

That said, others have modified OpenFAST to input motion instead of loads (e.g., see: https://github.com/OpenFAST/openfast/issues/617, https://github.com/OpenFAST/openfast/issues/795).

Moreover, other work has used OpenFAST to generate a digital twin that can be used to take measurements and reconstruct other responses that are not measured directly (e.g., see: https://www.nrel.gov/docs/fy20osti/76854.pdf).

Best regard,

**Comment by HiHoGit** [comment]
Thank you @jjonkman for the answer. I had a look at the paper and I have a couple of questions.
Is it possible to get the state-space matrices of the turbine when running a full non-linear module of OpenFAST or it must switch to the linearized module? 
it seems that when we want to run the linearized module, we have to set all controller settings to None. any possibility to have our own controller in this situation?

**Reply by jjonkman** [reply]
Dear @HiHoGit,

There is no "linearized module" of OpenFAST.  But the linearization functionality of OpenFAST can be used to extract the equivalent linear state-space model from the various OpenFAST modules and glue code.  The linearized models can be extracted at any point in time, but it is typically far more useful to linearize the model at specific (periodic) steady state operating points.

Other than some simple built-in control options, most controllers in OpenFAST are user-specified, and the OpenFAST linearization functionality is not set up to linearize user-specified controllers.  That said, once you've extracted a linear model from OpenFAST, it is likely not difficult to expand the linearized models to include linearized control logic in a post-processing step.

Best regards,

**Comment by HiHoGit** [comment]
Ok, Thank you for the clarification.
It means that there is no functionality of OpenFAST that can be used to extract the equivalent non-linear state-space model from the various OpenFAST modules and glue code. Is that correct?

Kind regards,

**Reply by jjonkman** [reply]
Dear @HiHoGit,

I would say that time-domain implementation of OpenFAST is a nonlinear state-space model.  But the nonlinear state-space model is not expressed as a single set of equations (that can be extracted from OpenFAST), rather, the equations are partitioned into separate modules, which are coupled through the glue code.  See our documentation on the OpenFAST modularization framework for more information: https://openfast.readthedocs.io/en/main/source/user/index.html#modularization-framework.

Best regards,

**Comment by HiHoGit** [comment]
Dear @jjonkman,

I have a question about OpenFAST linearization and using state-space matrices through a Kalman filter to estimate certain qualities. My governing equation is the simple tower first-order dynamic where I have the tower top fore-aft (FA) acceleration as a measurement (y = Cx + Du + w), and the thrust estimated by a Look-Up Table (LUT) (x' = Ax + Bu + v) as an input. I am attempting to estimate tower states and the tower base bending moment (z = Ex + Fu).

Here's what I've done: Initially, I performed linearization for a 2-degree-of-freedom (2DOF) system, obtained the ABCD matrices, and conducted the estimation. In a second approach, I introduced additional degrees of freedom (DOFs) for the tower and blades but kept the tower motion equation the same (2-DOF). I only utilized the associated values of tower stiffness and damping, eliminating other values in the matrices.

Following the same procedure and conditions, I observed that the estimation error of the tower top deflection did not change significantly (possibly due to consistent measurement and input). However, the estimation error of the tower base bending moment showed significant improvement.

I would appreciate it if you could provide any insight or explanation for this observation. I am somewhat concerned as I am unable to mathematically justify why I can eliminate certain values in the ABCD matrices and only consider what is necessary for the tower FA, even when the linearization is performed using 9DOF!

Thank you!

**Reply by jjonkman** [reply]
Dear @HiHoGit,

I'm not sure I fully understand enough about what you are doing to comment.

Perhaps our recently published paper on digital twins based on linearized OpenFAST models is useful for you: https://wes.copernicus.org/articles/9/1/2024/?

Perhaps @ebranlard can comment?

Best regards,

**Reply by HiHoGit** [reply]
Thanks a lot!

**Comment by HiHoGit** [comment]
Dear @ebranlard,

I have a question about OpenFAST linearization and using state-space matrices through a Kalman filter to estimate certain qualities. My governing equation is the simple tower first-order dynamic where I have the tower top fore-aft (FA) acceleration as a measurement (y = Cx + Du + w), and the thrust estimated by a Look-Up Table (LUT) (x' = Ax + Bu + v) as an input. I am attempting to estimate tower states and the tower base bending moment (z = Ex + Fu).

Here's what I've done: Initially, I performed linearization for a 2-degree-of-freedom (2DOF) system, obtained the ABCD matrices, and conducted the estimation. In a second approach, I introduced additional degrees of freedom (DOFs) for the tower and blades but kept the tower motion equation the same (2-DOF). I only utilized the associated values of tower stiffness and damping, eliminating other values in the matrices.

Following the same procedure and conditions, I observed that the estimation error of the tower top deflection did not change significantly (possibly due to consistent measurement and input). However, the estimation error of the tower base bending moment showed significant improvement.

I would appreciate it if you could provide any insight or explanation for this observation. I am somewhat concerned as I am unable to mathematically justify why I can eliminate certain values in the ABCD matrices (possibly due to the inclusion of thrust as an input) and only consider what is necessary for the tower FA, even when the linearization is performed using 9DOF! 

Thank you!
